File: XRViewer/AppStateController/AppStateController.m
#import "AppStateController.h"

#define RUN_ACTION_ASYNC_MAIN(a,p) if(a){ dispatch_async(dispatch_get_main_queue(), ^{ a(p);}); }

typedef void (^ExclusiveAction)(void);
typedef NS_ENUM(NSUInteger, ExclusiveStateType)
{
    ExclusiveStateMessage,
    ExclusiveStateMemory,
    ExclusiveStateBackground,
    ExclusiveStateReachbility
};


@interface ExclusiveState : NSObject

@property (nonatomic, copy) ExclusiveAction action;
@property (nonatomic, copy) NSString *url;
@property (nonatomic) ExclusiveStateType type;
@property (nonatomic) ShowMode mode;

@end


@implementation ExclusiveState
@end


@interface AppStateController ()

@property (nonatomic, strong) NSMutableArray *exclusives;

@end


@implementation AppStateController

- (instancetype)initWithState:(AppState *)state
{
    self = [super init];
    
    if (self)
    {
        [self setExclusives:[NSMutableArray new]];
        [self setState:state];
    }
    
    return self;
}

- (void)setShowMode:(ShowMode)mode
{
    if (mode != [[self state] showMode])
    {
        RUN_ACTION_ASYNC_MAIN([self onDebug], (mode == ShowMultiDebug));
    }
    
    [self setState:[[self state] updatedShowMode:mode]];
    
    RUN_ACTION_ASYNC_MAIN([self onModeUpdate], [[self state] showMode]);
}

- (void)setShowOptions:(ShowOptions)options
{
    [self setState:[[self state] updatedShowOptions:options]];
    
    RUN_ACTION_ASYNC_MAIN([self onOptionsUpdate], [[self state] showOptions]);
}

- (void)setRecordState:(RecordState)rState
{
    [self setState:[[self state] updatedRecordState:rState]];
    
    RUN_ACTION_ASYNC_MAIN([self onRecordUpdate], [[self state] recordState]);
    
    if (((rState == RecordStateRecordingWithMicrophone) && ([[self state] micEnabled] == NO)) ||
        ((rState == RecordStateRecording) && [[self state] micEnabled]))
    {
        [self invertMic];
    }
}

- (void)setWebXR:(BOOL)webXR
{
    [self setState:[[self state] updatedWebXR:webXR]];
    
    RUN_ACTION_ASYNC_MAIN([self onXRUpdate], [[self state] webXR]);
}

- (void)setARRequest:(NSDictionary *)dict
{
    [self setState:[[self state] updatedWithARRequest:dict]];
    
    RUN_ACTION_ASYNC_MAIN([self onRequestUpdate], dict);
}

- (void)setARInterruption:(BOOL)interruption
{
    [self setState:[[self state] updatedWithInterruption:interruption]];
    
    RUN_ACTION_ASYNC_MAIN([self onInterruption], interruption);
}

- (BOOL)shouldShowURLBar
{
    if ([[self state] recordState] == RecordStatePhoto ||
        [[self state] recordState] == RecordStateRecording ||
        [[self state] recordState] == RecordStateRecordingWithMicrophone ||
        [[self state] recordState] == RecordStateGoingToRecording ||
        (([[self state] recordState] == RecordStatePreviewing) && ([[UIDevice currentDevice] userInterfaceIdiom] != UIUserInterfaceIdiomPad)))
    {
        return NO;
    }
    
    BOOL showURLBar = NO;
    
    if ([[self state] webXR] == NO) {
        showURLBar = YES;
    } else {
        if([[self state] showMode] == ShowDebug) {
            showURLBar = NO;
        } else if ([[self state] showMode] == ShowMulti) {
            showURLBar = YES;
        } else if ([[self state] showMode] == ShowMultiDebug) {
            showURLBar = YES;
        }
    }
    
    return showURLBar;
}

- (BOOL)shouldSendARKData
{
    return [[self state] webXR] && [[self state] aRRequest];
}

- (BOOL)shouldSendCVData {
    return [[self state] computerVisionFrameRequested] && [[self state] sendComputerVisionData] && [[self state] userGrantedSendingComputerVisionData];
}

- (BOOL)shouldSendNativeTime {
    return [[self state] numberOfTimesSendNativeTimeWasCalled] < 9;
}

- (void)invertMic
{
    BOOL micEnabledUpdated = ![[self state] micEnabled];
    
    [self setState:[[self state] updatedWithMicEnabled:micEnabledUpdated]];
    
    RUN_ACTION_ASYNC_MAIN([self onMicUpdate], micEnabledUpdated);
}

- (void)invertShowMode
{
    [[self state] showMode] == ShowSingle ? [self setShowMode:ShowMulti] : [self setShowMode:ShowSingle];
}

- (void)invertDebugMode
{
    [[self state] showMode] == ShowMulti ? [self setShowMode:ShowMultiDebug] : [self setShowMode:ShowMulti];
}

- (BOOL)wasMemoryWarning
{
    __block BOOL was = NO;
    
    [[self exclusives] enumerateObjectsUsingBlock:^(ExclusiveState *  _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop)
     {
         if ([obj type] == ExclusiveStateMemory)
         {
             was = YES;
             *stop = YES;
         }
     }];
    
    return was;
}

- (BOOL)isRecording
{
    switch ([[self state] recordState])
    {
        case RecordStateGoingToRecording:
        case RecordStatePhoto:
        case RecordStateRecording:
        case RecordStateRecordingWithMicrophone:
            return YES;
            
        default:
            return NO;
    }
}

- (void)saveOnMessageShowMode
{
    [self saveOnType:ExclusiveStateMessage url:nil mode:[[self state] showMode]];
}

- (void)applyOnMessageShowMode
{
    [self applyOnType:ExclusiveStateMessage];
}

- (void)saveDidReceiveMemoryWarningOnURL:(NSString *)url
{
    [self saveOnType:ExclusiveStateMemory url:url mode:0];
}

- (void)applyOnDidReceiveMemoryAction
{
    [self applyOnType:ExclusiveStateMemory];
}

- (void)saveMoveToBackgroundOnURL:(NSString *)url
{
    [self setShowMode:ShowNothing];
    [self saveOnType:ExclusiveStateBackground url:url mode:0];
}

- (void)applyOnEnterForegroundAction
{
    [self applyOnType:ExclusiveStateBackground];
}

- (void)saveNotReachableOnURL:(NSString *)url
{
    [self saveOnType:ExclusiveStateReachbility url:url mode:0];
}

- (void)applyOnReachableAction
{
    [self applyOnType:ExclusiveStateReachbility];
}

#pragma mark Private

- (void)saveOnType:(ExclusiveStateType)type url:(NSString *)url mode:(ShowMode)mode
{
    ExclusiveState *state = [ExclusiveState new];
    if (url != nil)
    {
        [state setUrl:url];
    }
    else
    {
        [state setMode:mode];
    }
    [state setType:type];
    
    __weak typeof (self) blockSelf = self;
    //__weak typeof (ExclusiveState *) blockState = state;
    
    switch (type)
    {
        case ExclusiveStateMessage:
        {
            [state setAction:^
             {
                 [blockSelf setShowMode:mode];
             }];
            break;
        }
        case ExclusiveStateMemory:
        {
            [state setAction:^
             {
                 RUN_ACTION_ASYNC_MAIN([blockSelf onMemoryWarning], url);
             }];
            break;
        }
        case ExclusiveStateBackground:
        {
            [state setAction:^
             {
                 RUN_ACTION_ASYNC_MAIN([blockSelf onEnterForeground], url);
             }];
            break;
        }
        case ExclusiveStateReachbility:
        {
            [state setAction:^
             {
                 RUN_ACTION_ASYNC_MAIN([blockSelf onReachable], url);
             }];
            break;
        }
    }
    
    [[self exclusives] addObject:state];
}

- (void)applyOnType:(ExclusiveStateType)type
{
    __block ExclusiveState *message = nil;
    
    [[self exclusives] enumerateObjectsUsingBlock:^(ExclusiveState *  _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop)
     {
         if ([obj type] == type)
         {
             message = obj;
             *stop = YES;
         }
     }];
    
    if (message != nil)
    {
        [message action]();
        [[self exclusives] removeObject:message];
    }
}

@end

File: XRViewer/AppStateController/AppStateController.swift
import UIKit

/**
 An enum representing the state of the app UI at a given time
 
 - nothing: Shows the warning labels
 - debug: Shows the helper and build label, and the AR debug info
 - url: Shows the URL Bar
 - urlDebug: Shows the URL Bar and the AR debug info
 */
@objc enum ShowMode: Int {
    case nothing    = 0
    case debug      = 1
    case url        = 2
    case urlDebug   = 3
}

enum ExclusiveStateType: Int {
    case exclusiveStateMessage
    case exclusiveStateMemory
    case exclusiveStateBackground
    case exclusiveStateReachability
}

class AppStateController: NSObject {
    @objc var state: AppState
    var onModeUpdate: ((ShowMode) -> Void)?
    var onOptionsUpdate: ((ShowOptions) -> Void)?
    @objc var onXRUpdate: ((Bool) -> Void)?
    @objc var onRequestUpdate: (([AnyHashable : Any]?) -> Void)?
    @objc var onDebug: ((Bool) -> Void)?
    @objc var onMemoryWarning: ((String?) -> Void)?
    @objc var onEnterForeground: ((String?) -> Void)?
    @objc var onReachable: ((String?) -> Void)?
    private var exclusives: [ExclusiveState] = []

    @objc init(state: AppState) {
        self.state = state
        super.init()

        exclusives = [ExclusiveState]()
    }

    func setShowMode(_ mode: ShowMode) {
        if mode != state.showMode {
            guard let onDebug = onDebug else { return }
            DispatchQueue.main.async {
                onDebug(mode == ShowMode.urlDebug)
            }
        }

        self.state = state.updatedShowMode(mode)
        guard let onModeUpdate = onModeUpdate else { return }
        let showMode = state.showMode
        DispatchQueue.main.async {
            onModeUpdate(showMode)
        }
    }

    func setShowOptions(_ options: ShowOptions) {
        state = state.updatedShowOptions(options)
        guard let onOptionsUpdate = onOptionsUpdate else { return }
        let showOptions = state.showOptions
        DispatchQueue.main.async {
            onOptionsUpdate(showOptions)
        }
    }

    @objc func setWebXR(_ webXR: Bool) {
        state = state.updatedWebXR(webXR)
        guard let onXRUpdate = onXRUpdate else { return }
        let webXR = state.webXR
        DispatchQueue.main.async {
            onXRUpdate(webXR)
        }
    }

    @objc func setARRequest(_ dict: [AnyHashable : Any], completed: @escaping () -> ()) {
        state = state.updated(withARRequest: dict)
        guard let onRequestUpdate = onRequestUpdate else { return }
        DispatchQueue.main.async {
            onRequestUpdate(dict)
            completed()
        }
    }

    @objc func invertDebugMode() {
        state.showMode == ShowMode.url ? setShowMode(ShowMode.urlDebug) : setShowMode(ShowMode.url)
    }

    @objc func shouldShowURLBar() -> Bool {

        var showURLBar = false

        if !state.webXR {
            showURLBar = true
        } else {
            if state.showMode == .debug {
                showURLBar = false
            } else if state.showMode == .url {
                showURLBar = true
            } else if state.showMode == .urlDebug {
                showURLBar = true
            }
        }

        return showURLBar
    }

    @objc func shouldSendARKData() -> Bool {
        return state.webXR && !state.aRRequest.isEmpty
    }

    @objc func shouldSendCVData() -> Bool {
        return state.computerVisionFrameRequested && state.sendComputerVisionData && state.userGrantedSendingComputerVisionData
    }

    @objc func shouldSendNativeTime() -> Bool {
        return state.numberOfTimesSendNativeTimeWasCalled < 2
    }

    @objc func wasMemoryWarning() -> Bool {
        var was = false

        (exclusives as NSArray).enumerateObjects({ obj, idx, stop in
            if let obj = obj as? ExclusiveState {
                if obj.type == .exclusiveStateMemory {
                    was = true
                    stop.pointee = true
                }
            }
        })

        return was
    }
    // rf ?

    @objc func saveOnMessageShowMode() {
        save(on: .exclusiveStateMessage, url: nil, mode: state.showMode)
    }

    @objc func applyOnMessageShowMode() {
        apply(on: .exclusiveStateMessage)
    }

    @objc func saveDidReceiveMemoryWarning(onURL url: String?) {
        if let aMode = ShowMode(rawValue: 0) {
            save(on: .exclusiveStateMemory, url: url, mode: aMode)
        }
    }

    @objc func applyOnDidReceiveMemoryAction() {
        apply(on: .exclusiveStateMemory)
    }

    @objc func saveMoveToBackground(onURL url: String?) {
        setShowMode(ShowMode.nothing)
        if let aMode = ShowMode(rawValue: 0) {
            save(on: .exclusiveStateBackground, url: url, mode: aMode)
        }
    }

    @objc func applyOnEnterForegroundAction() {
        apply(on: .exclusiveStateBackground)
    }

    @objc func saveNotReachable(onURL url: String?) {
        if let aMode = ShowMode(rawValue: 0) {
            save(on: .exclusiveStateReachability, url: url, mode: aMode)
        }
    }

    @objc func applyOnReachableAction() {
        apply(on: .exclusiveStateReachability)
    }

// MARK: Private

    func save(on type: ExclusiveStateType, url: String?, mode: ShowMode) {
        let state = ExclusiveState()
        if url != nil {
            state.url = url ?? ""
        } else {
            state.mode = mode
        }
        state.type = type

        weak var blockSelf: AppStateController? = self

        switch type {
            case .exclusiveStateMessage:
                state.action = {
                    blockSelf?.setShowMode(mode)
                }
            case .exclusiveStateMemory:
                state.action = {
                    guard let onMemoryWarning = blockSelf?.onMemoryWarning else {
                        print("Unable to set onMemoryWarning")
                        return
                    }
                    DispatchQueue.main.async {
                        onMemoryWarning(url)
                    }
                }
            case .exclusiveStateBackground:
                state.action = {
                    guard let onEnterForeground = blockSelf?.onEnterForeground else {
                        print("Unable to set onEnterForeground")
                        return
                    }
                    DispatchQueue.main.async {
                        onEnterForeground(url)
                    }
                }
            case .exclusiveStateReachability:
                state.action = {
                    guard let onReachable = blockSelf?.onReachable else {
                        print("Unable to set onReachable")
                        return
                    }
                    DispatchQueue.main.async {
                        onReachable(url)
                    }
                }
        }

        exclusives.append(state)
    }

    func apply(on type: ExclusiveStateType) {
        var message: ExclusiveState? = nil

        (exclusives as NSArray).enumerateObjects({ obj, idx, stop in
            if let obj = obj as? ExclusiveState {
                if obj.type == type {
                    message = obj
                    stop.pointee = true
                }
            }
        })

        if let message: ExclusiveState = message {
            if let action = message.action {
                action()
            }
            exclusives.removeAll { $0 == message }
        }
    }
}

class ExclusiveState: NSObject {
    var action: (() -> Void)?
    var url = ""
    var type: ExclusiveStateType?
    var mode: ShowMode?
}
